### PR TITLE
Deprecate ref-maybe-const for array formal arguments

### PR DESCRIPTION
**Pull request by jabraham17**
Deprecates ref-maybe-const for array formal arguments. Compiler will warn when a ref-maybe-const formal is inferred to be `ref` in `cullOverReferences`. This will not warn for compiler generated functions or outer variables (when outlining nested functions). It is written in a way that exceptions can be removed to implement other ref-maybe-const deprecations.

Implements part of the decisions for https://github.com/chapel-lang/chapel/issues/21398#issuecomment-1654660481

## Summary of changes
- add explicit intents (either `ref` or `const`) where needed in modules and tests
  - some functions, the body was modified to remove the need for an explicit intent (eg `ref a = ...` to `const ref a = ...`)
- tweaks `compiler/passes/expandExternArrayCalls.cpp` to interpret intents and use `c_ptrConst`.
- added temporaries for lvalue errors due to #22816 
  - tweaks `compiler/passes/normalize.cpp` to fix lvalue error with `convertToExternalArray`. 
- rewrite `bulkAdd` for sparse domains to prefer codepaths that do not require `sort`
  - calling sort along any codepath requires the `inds` to be marked as `ref`
  - instead of taking a `preserveInds` boolean argument,  now there is a separate `bulkAddNoPreserveInds`
- change uses of `c_ptrTo` to `c_ptrToConst` where applicable

## New tests
- add `test/deprecated/ref-maybe-const.chpl`
  - also added deprecation tests for this intents, task intents, and forall intents, these are marked notest for now
- futurize `test/functions/ferguson/ref-pair/ifexpr/ifexpr2-1.chpl`
- add `test/users/ibertolacc/void_extern_proc_array_ref.chpl`

## Testing
- paratest (llvm) with futures
- paratest (llvm) with gasnet
- paratest (c)
- paratest with --verify

[Reviewed by @vasslitvinov]

closes Cray/chapel-private#5218

### CODE REVIEW

File: test/arrays/ferguson/choose-array/choose-array3.chpl [hunk 1]
```diff
@@ -5,10 +5,10 @@ proc chooseArray(A:[] int, B:[] int, chooseA:bool)
    else return B;
 }
 
-proc updateAndPrint(A:[], B:[], C:[], D:[])
+proc updateAndPrint(ref A:[], ref B:[], C:[], D:[])
 {
-  C[1] = 8;
-  D[1] = 9;
+  A[1] = 8;
+  B[1] = 9;
```
Comment (vasslitvinov) [inline test/arrays/ferguson/choose-array/choose-array3.chpl:11]: I wonder if a better conversion would be to keep A, B const and instead give C, D the [non-const] in-intent.

File: test/arrays/ferguson/semantic-examples/1.5-return-alias-slice-global.chpl [hunk 2]
```diff
@@ -8,15 +8,15 @@ proc f() {
   return slice;
 }
 
-proc g(x) {
+proc g(in x) {
   x[2] = one;
 }
```
Comment (vasslitvinov) [inline test/arrays/ferguson/semantic-examples/1.5-return-alias-slice-global.chpl:13]: I think the switch to in-intent takes us away from answering the question at the bottom of this test. How about er replace `g(f())` with:
```chpl
f()[2] = one;
```

P.S. I am puzzled that my replacement works. However I only tried it on main, not sure whether it will work on your branch.

If we cannot make this and the subsequent tests answer the question they aim to, I do not see the value of keeping them.

Maybe check with @mppf ?

File: test/arrays/ferguson/semantic-examples/2-call-modifies-returned-array.chpl [hunk 3]
```diff
@@ -7,14 +7,15 @@ proc returnGlobalArray() {
...
+proc setToTwo(ref A) {
   A = ArrTwo;
 }
 
 writeln("globalArray is ", globalArray);
 writeln("running setToTwo(returnGlobalArray())");
 
-setToTwo(returnGlobalArray());
+var t = returnGlobalArray();
+setToTwo(t);
```
Comment (vasslitvinov) [inline test/arrays/ferguson/semantic-examples/2-call-modifies-returned-array.chpl:18]: The same comment as above.

File: test/arrays/ferguson/semantic-examples/4-return-alias-global.chpl [hunk 4]
```diff
@@ -8,15 +8,16 @@ proc f() {
...
 }
 
-proc g(x) {
+proc g(ref x) {
   x[2] = one;
 }
 
-g(f());
+var t = f();
+g(t);
```
Comment (vasslitvinov) [inline test/arrays/ferguson/semantic-examples/4-return-alias-global.chpl:16]: Ditto.

File: test/release/examples/primers/arrays.chpl [hunk 5]
```diff
@@ -157,9 +157,10 @@ proc negateAndPrintArr(X: [?D] real) {
 negateAndPrintArr(B);
 
 //
-// Arrays are passed to routines by reference (``ref``) by default, so
-// the modifications to ``X`` in procedure ``negateAndPrintArr()`` are
-// reflected back in the actual argument ``B`` as well:
+// Arrays are passed to routines by constant reference (``const ref``) by
+// default, so for the modification to ``X`` in procedure
+// ``negateAndPrintArr()`` to be legal and reflected back to the actual argument
+// ``B``  the explicit ``ref`` must be used.
```
Comment (vasslitvinov) [inline test/release/examples/primers/arrays.chpl:163]: I propose:

```
// Arrays are passed to a routine by constant reference (``const ref``) by
// default, which does not allow them to be modified within the routine.
// Therefore the above procedure ``negateAndPrintArr()`` uses a non-constant
// reference intent ``ref`` explicitly, so that its modifications of ``X``
// are allowed and reflected back in the actual argument ``B``:
```

File: test/studies/cholesky/jglewis/version2/elemental/scalar_inner_product_cholesky.chpl [hunk 6]
```diff
@@ -83,7 +83,7 @@ module scalar_inner_product_cholesky  {
 	A (j, j)      = sqrt ( A (j, j) );
 	A (j+1.., j ) = A (j+1.., j) / A (j, j);
       }
-      else 
+      else
```
Comment (vasslitvinov) [inline test/studies/cholesky/jglewis/version2/elemental/scalar_inner_product_cholesky.chpl:86]: In these cholesky tests, whitespace changes also outshadow the PR-related changes. It would be nicer without them.

File: test/studies/cholesky/jglewis/version2/elemental/scalar_inner_product_cholesky.chpl [hunk 7]
```diff
@@ -92,4 +92,3 @@ module scalar_inner_product_cholesky  {
 
   }
 }
-
```
Comment (vasslitvinov) [inline test/studies/cholesky/jglewis/version2/elemental/scalar_inner_product_cholesky.chpl:95]: I think you eliminated "no newline at end of file". If this is something else, please double-check.

File: test/studies/cholesky/marybeth/chol.chpl [hunk 8]
```diff
@@ -27,12 +27,12 @@ proc main() {
 
   chol(A);
 
-  writeln("Factored Matrix:");
+writeln("Factored Matrix:");
```
Comment (vasslitvinov) [inline test/studies/cholesky/marybeth/chol.chpl:30]: Whitespace deleted accidentally. No big deal however.

File: test/studies/graph500/kristyn/Graph500_1D_onV/Graph500_Modules/Verify.chpl [hunk 9]
```diff
@@ -139,19 +139,19 @@ proc compute_levels (ParentTree, Level): int
...
          while ( Level[parent] < 0) {
            Level[parent] = nhop;
-           nhop -= 1; 
+           nhop -= 1;
            parent = ParentTree[parent];
          }
        }
 
-     } 
+     }
```
Comment (vasslitvinov) [inline test/studies/graph500/kristyn/Graph500_1D_onV/Graph500_Modules/Verify.chpl:154]: Whitespace ditto in test/studies/graph500

Would be much nicer in a separate PR.

File: test/users/rbarrett/diffuse2d-v3-blc.chpl [hunk 10]
```diff
@@ -410,29 +410,29 @@ proc ApplyStencil3d27pt ( ProblemSpace2d, newGrid, oldGrid, iteration ) {
...
+
    var
       err : real;  // Computed error.
 
    /* --------------------- */
    /* Executable Statements */
    /* --------------------- */
 
-   if ( verbose ) 
+   if ( verbose )
```
Comment (vasslitvinov) [inline test/users/rbarrett/diffuse2d-v3-blc.chpl:435]: Here is another example where there are 4 significant changes and 80+ (?) whitespace adjustments.

File: test/release/examples/primers/learnChapelInYMinutes.chpl [hunk 11]
```diff
@@ -497,7 +497,7 @@ Procedures
 ----------
 */
 
-// Chapel procedures have similar syntax functions in other languages. 
+// Chapel procedures have similar syntax functions in other languages.
```
Comment (vasslitvinov) [inline test/release/examples/primers/learnChapelInYMinutes.chpl:500]: While there, let us fix it "similar syntax TO functions ..." or somesuch.

**Review by vasslitvinov — APPROVED** [review]
For the OP / PR message:

* array argument formals --> array formal arguments [my preference]

* a ref-maybe-const formal is marked as ref:
  "is marked as" --> "is inferred to be" ?

* Please explain your addition of dsiBulkAddNoPreserveInds().
  It would also be good to add a comment in the code explaining what it does.

For the cases where you switched the default intent to `const`, ex. in ChapelArray.chpl or in tests like `test/types/tuple/ferguson/tuple-array-concrete.chpl`, I am curious why not `ref` and/or why an explicit intent was needed. I suspect the compiler wants to infer it as `ref`? why?

In `test/types/tuple/ferguson/tuple-array.chpl` and siblings, how well do you think your changes there preserve the intent of those tests?

File: test/arrays/ferguson/semantic-examples/1.5-return-alias-slice-global.chpl (hunk 2 above)
Comment (jabraham17) [reply test/arrays/ferguson/semantic-examples/1.5-return-alias-slice-global.chpl:13]: I think a temporary `ref` variable for the result of `f` will best preserve the intentions of the test

Comment (jabraham17) [reply test/arrays/ferguson/semantic-examples/1.5-return-alias-slice-global.chpl:13]: Thinking about these more, I think `var` actually does capture it best, since `f` is returning by value. The reason these temporaries are added is that there are lvalue errors since we changes how that works for arrays recently, which I think has a bad reaction with ref-maybe-const

File: test/users/franzf/v2/chpl/fft_512.chpl [hunk 12]
```diff
@@ -268,17 +268,17 @@ proc fft512(Y: [] complex, X: [] complex) {
...
+    s3515, s3516, s3517, s3518, s3519, s3520, s3521, s3522,
+    s3523, s3524, t5792, t5793, t5794, t5795, t5796, t5797,
+    t5798, t5799, t5800, t5801, t5802, t5803, t5804, t5805,
+    t5806, t5807, t5808, t5809, t5810, t5811, t5812, t5813,
+    t5814, t5815, t5816, t5817, t5818, t5819, t5820, t5821,
     t5822:complex;
-        var a3500, a3501, a3502, a3503, a3504, a3505, a3506, 
-    a3507, a3508, a3509, a3510, a3511, a3512, a3513, a3514, 
+        var a3500, a3501, a3502, a3503, a3504, a3505, a3506,
+    a3507, a3508, a3509, a3510, a3511, a3512, a3513, a3514,
```
[the remark this answers is not in the document]
Comment (jabraham17) [reply test/users/franzf/v2/chpl/fft_512.chpl:281]: All of the "just trailing spaces" changes are auto-fixes by my editor. I think it lightens up my PR diff, but increases the quality of some of our old code prior to the pre-merge spaces check

**Comment by vasslitvinov** [comment]
> All of the "just trailing spaces" changes are auto-fixes by my editor. I think it lightens up my PR diff, but increases the quality of some of our old code prior to the pre-merge spaces check

The space check does not apply to tests. Also, to me trailing space has no impact on the quality of code.

Excessive trailing space changes affect the PR review experience, however, and are best done in a separate PR.

File: test/release/examples/primers/arrays.chpl (hunk 5 above)
Comment (jabraham17) [reply test/release/examples/primers/arrays.chpl:163]: I have changed it to the following, which is slightly tweaked:

```chapel
// Arrays are passed to routines by constant reference (``const ref``) by
// default, which does not allow them to be modified within the routine.
// The above procedure ``negateAndPrintArr()`` must use a non-constant
// reference intent (``ref``) explicitly, so that its modifications of ``X``
// are both allowed and reflected in the actual argument ``B``:
```

**Comment by jabraham17** [comment]
>For the cases where you switched the default intent to const, ex. in ChapelArray.chpl or in tests like test/types/tuple/ferguson/tuple-array-concrete.chpl, I am curious why not ref and/or why an explicit intent was needed. I suspect the compiler wants to infer it as ref? why?
>
> In test/types/tuple/ferguson/tuple-array.chpl and siblings, how well do you think your changes there preserve the intent of those tests?

These were both changes I made with @mppf, where the rationale was we are ok forcing users to explicitly mark varargs as `const`. In `ChapelArray`, it was added to additional places other than just varargs so the intents of overloaded functions matched

**Comment by jabraham17** [comment]
@vasslitvinov I think I have implemented all of your feedback except the spaces. I will look at separating those out after I resolve the merge conflicts, but I think for this PR it may be more trouble than it is worth at this point

**Comment by bradcray** [comment]
Woohoo!
